hal/src/rtl872x/exflash_hal.cpp

```diff
@@ -291,36 +595,61 @@ static bool is_block_erased(uintptr_t addr, size_t size) {
 }
 
 __attribute__((section(".ram.text"), noinline))
-static int erase_common(uintptr_t start_addr, size_t num_blocks, int len) {
-    ExFlashLock lk;
-
-    const size_t block_length = len == EraseSector ? 4096 : 64 * 1024;
-
-    /* Round down to the nearest block */
-    start_addr = ((start_addr / block_length) * block_length);
-
-    for (size_t i = 0; i < num_blocks; i++) {
-        if (!is_block_erased(start_addr + i * block_length, block_length)) {
-            FLASH_Erase(len, start_addr + i * block_length);
-        }
+static int erase_common(uintptr_t address, hal_exflash_cmd_t cmd) {
+    ExFlashLock lk(false);
+    size_t size = 0;
+    if (cmd == GENERIC_FLASH_CMD_SE) {
+        size = 4 * 1024;
+    } else if (cmd == GENERIC_FLASH_CMD_BE32K) {
+        size = 32 * 1024;
+    } else if (cmd == GENERIC_FLASH_CMD_BE64K) {
+        size = 64 * 1024;
+    } else {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
     }
+    address = ((address / (size)) * (size));
+    if (is_block_erased(address, (size))) {
+        return size;
+    }
+    RtkSpicFlashConfigGuard guard;
 
-    // LOG_DEBUG(ERROR, "Erased %lu %lukB blocks starting from %" PRIxPTR,
-    //           num_blocks, block_length / 1024, start_addr);
-
-    dcacheCleanInvalidateAligned(SPI_FLASH_BASE + start_addr, block_length * num_blocks);
-
-    return SYSTEM_ERROR_NONE;
+    rtkSpiFlashSendCmd(GENERIC_FLASH_CMD_WREN);
+    /* Disable SPI_FLASH */
+    SPIC->ssienr = 0;
+    rtkSpiSetTxMode();
+    /* set flash cmd + addr and write to fifo */
+    SPIC->dr[0].word = RTK_SPIC_CMD_ADDR_FORMAT(cmd, address);
+    rtkSpiFlashEnableCsWriteErase();
+    rtkSpiFlashSendCmd(GENERIC_FLASH_CMD_WRDI);
+    return size;
 }
 
 __attribute__((section(".ram.text"), noinline))
 int hal_exflash_erase_sector(uintptr_t start_addr, size_t num_sectors) {
-    return erase_common(start_addr, num_sectors, EraseSector);
+    size_t sizeInBytes = num_sectors * 4096;
+    uintptr_t addr = start_addr;
+    while (sizeInBytes > 0) {
+        hal_exflash_cmd_t cmd = GENERIC_FLASH_CMD_SE;
+        if (addr % (64*1024) == 0 && sizeInBytes >= 64*1024) {
+            cmd = GENERIC_FLASH_CMD_BE64K;
+        } else if (addr % (32*1024) == 0 && sizeInBytes >= 32*1024) {
+            cmd = GENERIC_FLASH_CMD_BE32K;
+        }
+        size_t size = CHECK(erase_common(addr, cmd));
+        addr += size;
+        sizeInBytes -= size;
+    }
+    dcacheCleanInvalidateAligned(SPI_FLASH_BASE + start_addr, num_sectors * 4096);
+    return 0;
 }
 
 __attribute__((section(".ram.text"), noinline))
 int hal_exflash_erase_block(uintptr_t start_addr, size_t num_blocks) {
-    return erase_common(start_addr, num_blocks, EraseBlock);
+    for (size_t i = 0; i < num_blocks; i++) {
+        CHECK(erase_common(start_addr, GENERIC_FLASH_CMD_BE64K));
```

Comment:
Do we actually do 64k block erases at some point in normal operation?

Reply:
Not really, no. But now we will be using 32K and 64K block erasures during at least OTA as `hal_exflash_erase_sector` optimizes this.